text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,299 @@
+
+1. If any requested mixin is not provided by the stack, the lifecycle will compare the missing mixins to the static list of mixins provided by stack buildpacks. If any mixins are still not provided, the build will fail.
+1. The lifecycle will run the detect phase for all stackpacks defined in the builder.
+1. The lifecycle will execute the stack buildpack build phase for all passing stackpack(s). If no stack buildpacks pass detect, the build will continue the build phase as normal (running userspace buildpacks).
+1. After, during, or before the lifecycle's build phase, the lifecycle will begin a new phase called _extend_.
+1. During the extend phase, stack buildpacks that passed detection will run against the run images accordingly (see details below).
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+ A stack buildpack (a.k.a. stackpack) is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* Does not have access to the application source code
+* Is run before all regular buildpacks
+* Is run against both the build and run images
+* Is distributed with the stack run and/or build image
+* May not create layers using the `<layers>` directory
+
+The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.
+
+For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase excluding the a few specific directories and files.
+
+Additionally, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory. By default, an implicit order will be created for all stackpacks included in a stack. The order can be overridden in the `builder.toml` with the following configuration:
+
+```
+[[stack.order]]
+  [[stack.order.group]]
+    id = "<stackpack id>"
+```
+
+A stackpack will only execute if it passes detection. When the stackpack is executed, its detect and build scripts use the same parameters as the regular buildpacks. But the arguments may differ. For example, the first positional argument to the `bin/build` (the `<layers>` directory) MUST NOT be a standard layers directory adhering to the [Buildpacks Build specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#build).
```

Comment:
I find this description of the input confusing.

Should we say that the `/bin/build` script will have the same arity as the regular buildpacks instead of the same parameters? Maybe we should just describe the expected inputs to the stack buildpack without comparing them to a regular buildpack? "The `<layers>` dir must not be the layers dir" (paraphrasing) is a confusing way to think about thing imho.